test/Microsoft.ML.Functional.Tests/Debugging.cs

```diff
@@ -0,0 +1,214 @@
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using Microsoft.ML.Data;
+using Microsoft.ML.Functional.Tests.Datasets;
+using Microsoft.ML.RunTests;
+using Microsoft.ML.TestFramework;
+using Microsoft.ML.Trainers;
+using Microsoft.ML.Transforms.Text;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace Microsoft.ML.Functional.Tests
+{
+    public class Debugging : BaseTestClass
+    {
+        public Debugging(ITestOutputHelper output) : base(output)
+        {
+        }
+
+        /// <summary>
+        /// Debugging: The individual pipeline steps can be inspected to see what is happening to 
+        /// data as it flows through.
+        /// </summary>
+        /// <remarks>
+        /// It should, possibly through the debugger, be not such a pain to actually
+        /// see what is happening to your data when you apply this or that transform. For example, if I
+        /// were to have the text "Help I'm a bug!" I should be able to see the steps where it is
+        /// normalized to "help i'm a bug" then tokenized into ["help", "i'm", "a", "bug"] then
+        /// mapped into term numbers [203, 25, 3, 511] then projected into the sparse
+        /// float vector {3:1, 25:1, 203:1, 511:1}, etc. etc.
+        /// </remarks>
+        [Fact]
+        void InspectIntermediatePipelineSteps()
+        {
+            var mlContext = new MLContext(seed: 1);
+
+            var data = mlContext.Data.LoadFromEnumerable<TweetSentiment>(
+                new TweetSentiment[]
+                {
+                    new TweetSentiment { Sentiment = true, SentimentText = "I love ML.NET." },
+                    new TweetSentiment { Sentiment = true, SentimentText = "I love TLC." },
+                    new TweetSentiment { Sentiment = false, SentimentText = "I dislike fika." }
+                });
+
+            // create a training pipeline.
+            var pipeline = mlContext.Transforms.Text.FeaturizeText(
+                "Features",
+                new TextFeaturizingEstimator.Options
+                {
+                    KeepPunctuations = false,
+                    OutputTokens = true,
+                    CharFeatureExtractor = null, // new WordBagEstimator.Options { NgramLength = 0, SkipLength = -1 },
+                    WordFeatureExtractor = new WordBagEstimator.Options { NgramLength = 1},
+                    Norm = TextFeaturizingEstimator.NormFunction.None
+                },
+                "SentimentText");
+
+            // Fit the pipeline to the data.
+            var model = pipeline.Fit(data);
+
+            // Transform the data.
+            var transformedData = model.Transform(data);
+
+            var preview = transformedData.Preview();
+
+            // Verify that columns can be inspected.
+            // Validate the tokens column.
+            var tokensColumn = transformedData.GetColumn<string[]>(transformedData.Schema["Features_TransformedText"]);
```

Comment:
>Features_TransformedText [](start = 91, length = 24)

where is this magic string coming from?

Reply:
This magic string is the tokenized column. It takes the name you give it, and returns `${OutputColumnName}_TransformedText}` I'll file a separate issue on it.

---
In reply to: [265701391](https://github.com/dotnet/machinelearning/pull/2937#discussion_r265701391) [](ancestors = 265701391)

Reply:
Issue #2957

---
In reply to: [265708837](https://github.com/dotnet/machinelearning/pull/2937#discussion_r265708837) [](ancestors = 265708837,265701391)